README.rst
========
Python Power Service and Client
========

Package generated to interface with a trading system. Augmented with a prototypical client

Package will return a random set of open trades for a given day, where each trade contains data for hours, trading volume, and an unique ID.
Client will process the data aggregating them on an hourly base after data validation and QC

Note: The package will naively return series for a given date and not take into account for example if the date is in the future.
Further the numbers for the volume are random and do not have any relationship with previous or subsequent numbers as would normally be the case in real data.

Installation
============
Through a terminal navigate to the folder you have the powerservice folder and run

::

    pip install ./python-powerservice


Documentation
=============

The service will be part of the python environment and can be called in code
::

    from powerservice import trading

Example that will output some trades
::
    from powerservice import trading
    trades = trading.get_trades("29/06/2021")

    print(trades)

Example that will process some trades
::
    from powerservice import trading
    from powerservice import client
    trade_date = "29/06/2021"
    output_path = "/tmp/"

    trades = trading.get_trades(trade_date)

    # Configure and instantiate the processor
    processor = PetroineosChallenge(
        client.DataValidator(),
        client.MapReduce(),
        client.PersistenceUnit(output_path))

    # process data
    processor.process(trades, trade_date)

    # NOTE: The data are persisted according to the persistence unit injected
    # In this case they are saved on the filesystem
    # in the location specified by  output_path




Excercise
=========

The exercise requires to aggregate on hourly basis the data volumes provided by timeseries generator on a 5 minutes tick.
Additionally we are required to provide a data profiling and a data quality report.

Considerations
--------------
Without a properly defined business domain context, the requirements provided are not sufficient to
fully qualify the problem.
In a real worlds setup one would consider contacting the stakeholders for clarifications.
In this exercise it is not possible to interact with the stakeholders.
Therefore, one is forced to make some assumptions and justify them.

Input Data Structure
--------------------
The generator provides a list containing a random number of timeseries.
According to the identifier we can assume each time-series relates to the same trade.
Moreover, it is not clear what instrument does each time-series represent,
thus if they can be added together or not.
For sake of exercise we would consider those time-series homogeneous and handle them
as if they were all trades for the same instrument.
Extra to that, having so may data point is quite unusual for a single trade in the most common asset classes.
I.e. in a real world setup one might expect a time-series to be built of many trades with different IDs per trade
Possibly the same id mapping to a small number of  different events e.g. in case of partial fills.
Given there are no strong conditions on the id, we simply ignore the case and assume the volume needs to be aggregated
across all valid events in a time interval and across all "trades".
The example provided in the Additional Notes section shows the aggregate volume to be either 150 till 9:00 or 80
after 9:00am. The expected output is not consisten with the example input
Therefore, one should simply assume the volume data need to be added.

Null values
-----------
The generator deliberately sets 1 in 16 time values to NaN, knowing the generating model
allows to interpolate on the timeline and reinstate the missing values.
However, In a real world setup, this condition is not granted and missing values require to be handled adequately.
For sake of this exercise we'll report the row as a data quality issue and mark it for reconciliation
TO note, this approach results in the volume of trades for the  for the day to underestimate the volumes potentially traded as those rows are discarded
We cannot make any strong assertion about the trades aggregated on an hourly basis as we won't be able to impute any of those trades to a particular time slot
The user should consider discarded rows and validate the volumes on a case by case basis.
In real life a datasource with such big data quality issues should be handled with care and assess if it is suitable for purpose

Naming Conventions
------------------
The naming of the files is somehow confusing.
The Filename pattern hints at the fact we'll have a file per hour.
However, the example in the additional notes shows a file per day.
Given the situation we'll produce a file per date with the time part set to teh midnight GMT mapped to london time locale

Timezone Handling
-----------------
The Time Zone requirement is difficult to understand the phrasing is not clear and the description is conflicting
with the examples provided. The original phrasing below

```
All values must be aggregated per hour (local/wall clock time). Note that for a given day, the
actual local start time of the day is 23:00 (11 pm) on the previous day. Local time is London
Local time for the day.
```

should  be interpreted consistently with the usual issue of mapping
timezones on utc time e.g. midnight on British summer time maps to 11pm of the previous day in UTC
Therefore, one can interpret the requirement as

```
The API returns time in the Europe/London timezone,  the output should be returned in UTC timezone.
Therefore during British Summer Time the values returned by the API for midnight is mapped to 23:00 of the previous day
```

Implementation
==============

The implementation is quite trivial. Once the specs are clarified it reduces to a simple ETL process.
One can use a bridge pattern (https://en.wikipedia.org/wiki/Bridge_pattern) with an outer class acting as orchestrator and few handlers implementing the processing steps injected to the constructor.

This approach has multiple advantages. Firts of all the orchestrator is agnostic to the actual implementation of the processing steps.
Moreover, each processing step is self contained and loosely coupled with the rest of the system.
This approach keeps every component very lean,
each component can be tested in isolation,
components can be easily replaced with minor impact on the codebase.
e.g. if we want to persist in a database we just need to inject a new persistence unit writing to a DB.
The same happens on the validation, e.g. if we want to add or modify the validity constraints we can inject a different validator
and it will do the job seamlessly (even filtering for valid rows is carried out by the validator instance using inversion of control)




CHANGELOG.rst

Changelog
=========

0.0.0 (2021-06-24)
------------------
Original Implementation of teh mock timeseries generator


0.1.0 (2022-09-15)
------------------
Added a client to the generator



tests/test_mapreduce.py
""" Tests for the MapReduce"""
from pandas.testing import assert_frame_equal

from powerservice.client import MapReduce
from powerservice.utils_for_testing import str_to_df


def test_map_reduce():
    input = str_to_df("""
        date______  time_   volume  id______________________________
    0   01/08/2022	00:00   100     07d01a7a08c646a68a3cc71b72c337c8
    1   01/08/2022	00:05   200     07d01a7a08c646a68a3cc71b72c337c8
    2   01/08/2022	00:10   300     07d01a7a08c646a68a3cc71b72c337c8
    4   01/08/2022	01:00   500     07d01a7a08c646a68a3cc71b72c337c8
    5   01/08/2022	01:05   600     07d01a7a08c646a68a3cc71b72c337c8
    6   01/08/2022	01:10   700     07d01a7a08c646a68a3cc71b72c337c8
    7   01/08/2022	00:00   110     07d01a7a08c646a68a3cc71b72c337c9
    8   01/08/2022	00:05   220     07d01a7a08c646a68a3cc71b72c337c9
    9   01/08/2022	00:10   330     07d01a7a08c646a68a3cc71b72c337c9
    """)

    expected = str_to_df("""
~~~~~~~~Local Time~~Volume
~~~~0~~~23:00~~~~~~~1260
~~~~1~~~00:00~~~~~~~1800
    """, delimiter='~')

    map_reduce = MapReduce()
    actual = map_reduce.map_reduce(input)
    assert_frame_equal(actual, expected)


tests/test_validator.py
""" Tests for the DataValidator"""
from pandas.testing import assert_frame_equal
from powerservice.client import DataValidator
from powerservice.utils_for_testing import str_to_df


def test__check_invalid_time_format():

    input = str_to_df("""
        date______  time_   volume  id
    0   01/08/2022	00:00   312     07d01a7a08c646a68a3cc71b72c337c8
    1   01/08/2022	        312     07d01a7a08c646a68a3cc71b72c337c8
    2   01/08/2022	xy:qw   312     07d01a7a08c646a68a3cc71b72c337c8
    """)

    expected = str_to_df("""
        date______  time_   volume  id______________________________  invalid_time_format
    0   01/08/2022	00:00   312     07d01a7a08c646a68a3cc71b72c337c8  False
    1   01/08/2022	        312     07d01a7a08c646a68a3cc71b72c337c8  True
    2   01/08/2022	xy:qw   312     07d01a7a08c646a68a3cc71b72c337c8  True
    """)

    data_validator = DataValidator()
    actual = data_validator._check_invalid_time_format(input)
    assert_frame_equal(actual, expected)

def test__check_invalid_volume():
    input = str_to_df("""
        date______  time_   volume  id______________________________
    0   01/08/2022	00:00   312     07d01a7a08c646a68a3cc71b72c337c8
    1   01/08/2022	00:05           07d01a7a08c646a68a3cc71b72c337c8
    2   01/08/2022	00:10   xyz     07d01a7a08c646a68a3cc71b72c337c8
    """)

    expected = str_to_df("""
        date______  time_   volume  id______________________________  invalid_volume
    0   01/08/2022	00:00   312     07d01a7a08c646a68a3cc71b72c337c8  False
    1   01/08/2022	00:05           07d01a7a08c646a68a3cc71b72c337c8  True
    2   01/08/2022	00:10   xyz     07d01a7a08c646a68a3cc71b72c337c8  True
    """)

    data_validator = DataValidator()
    actual = data_validator._check_invalid_volume(input)
    assert_frame_equal(actual, expected)



def test__check_unexpected_time():
    input = str_to_df("""
        date______  time_   volume  id______________________________
    0   01/08/2022	00:00   312     07d01a7a08c646a68a3cc71b72c337c8
    1   01/08/2022	00:01   123     07d01a7a08c646a68a3cc71b72c337c8
    2   01/08/2022	25:10   123     07d01a7a08c646a68a3cc71b72c337c8
    """)

    expected = str_to_df("""
        date______  time_   volume  id______________________________  unexpected_time
    0   01/08/2022	00:00   312     07d01a7a08c646a68a3cc71b72c337c8  False
    1   01/08/2022	00:01   123     07d01a7a08c646a68a3cc71b72c337c8  True
    2   01/08/2022	25:10   123     07d01a7a08c646a68a3cc71b72c337c8  True
    """)

    data_validator = DataValidator()
    actual = data_validator._check_unexpected_time(input)
    assert_frame_equal(actual, expected)

def test__check_missing_time():
    # Test omitted as it is too verbose for the purpose of this exercise
    # in production setup one can define and designate some predefined files for input and expected results
    pass


def test_get_valid_trades():
    input = str_to_df("""
            date______  time_   volume  id______________________________  invalid_time_format  invalid_volume  unexpected_time  missing_time
        0   01/08/2022	00:00   312     07d01a7a08c646a68a3cc71b72c337c8  False                False           False            False
        1   01/08/2022	25:05   123     07d01a7a08c646a68a3cc71b72c337c8  True                 False           False            False
        2   01/08/2022	00:10           07d01a7a08c646a68a3cc71b72c337c8  False                True            False            False
        3   01/08/2022	00:01   123     07d01a7a08c646a68a3cc71b72c337c8  False                False           True             False
        """)
    expected = str_to_df("""
            date______  time_   volume  id______________________________  invalid_time_format  invalid_volume  unexpected_time  missing_time
        0   01/08/2022	00:00   312.    07d01a7a08c646a68a3cc71b72c337c8  False                False           False            False
        """)

    data_validator = DataValidator()
    actual = data_validator.get_valid_trades(input)
    assert_frame_equal(actual, expected)


def test_get_trade_exceptions():
    input = str_to_df("""
            date______  time_   volume  id______________________________  invalid_time_format  invalid_volume  unexpected_time  missing_time
        0   01/08/2022	00:00   312     07d01a7a08c646a68a3cc71b72c337c8  False                False           False            False
        1   01/08/2022	25:05   123     07d01a7a08c646a68a3cc71b72c337c8  True                 False           False            False
        2   01/08/2022	00:10           07d01a7a08c646a68a3cc71b72c337c8  False                True            False            False
        3   01/08/2022	00:01   123     07d01a7a08c646a68a3cc71b72c337c8  False                False           True             False
        """)
    expected = str_to_df("""
            date______  time_   volume  id______________________________  invalid_time_format  invalid_volume  unexpected_time  missing_time
        1   01/08/2022	25:05   123     07d01a7a08c646a68a3cc71b72c337c8  True                 False           False            False
        2   01/08/2022	00:10           07d01a7a08c646a68a3cc71b72c337c8  False                True            False            False
        3   01/08/2022	00:01   123     07d01a7a08c646a68a3cc71b72c337c8  False                False           True             False
        """)

    data_validator = DataValidator()
    actual = data_validator.get_trade_exceptions(input)
    assert_frame_equal(actual, expected)


def test_get_data_quality_summary():
    input = str_to_df("""
            date______  time_   volume  id______________________________  invalid_time_format  invalid_volume  unexpected_time  missing_time
        1   01/08/2022	25:05   123     07d01a7a08c646a68a3cc71b72c337c8  True                 False           False            False
        2   01/08/2022	00:10           07d01a7a08c646a68a3cc71b72c337c8  False                True            False            False
        3   01/08/2022	00:01   123     07d01a7a08c646a68a3cc71b72c337c8  False                False           True             False
        4   01/08/2022	00:02   123     07d01a7a08c646a68a3cc71b72c337c8  False                False           True             False
        5   01/08/2022	00:10   123     07d01a7a08c646a68a3cc71b72c337c8  False                False           False            False
        """)
    expected = str_to_df("""
        id                                missing_time_count  invalid_time_format_count  unexpected_time_count  invalid_volume_count  time_min  time_max
        07d01a7a08c646a68a3cc71b72c337c8  0                   1                          2                      1                     00:10     00:10
        """)

    data_validator = DataValidator()
    actual = data_validator.get_data_quality_summary(input)
    assert_frame_equal(actual, expected)


tests/test_persistence_unit.py
""" Tests for the MapReduce"""
from unittest.mock import MagicMock

import pandas as pd

from powerservice.client import PersistenceUnit


def test_save_results():
    trade_date_str = "20220801"
    trade_time_str = "0000"
    aggregated_data = pd.DataFrame()
    aggregated_data.to_csv = MagicMock()

    trade_exceptions_df = pd.DataFrame()
    trade_exceptions_df.to_csv = MagicMock()

    data_quality_summary = pd.DataFrame()
    data_quality_summary.to_csv = MagicMock()

    persistence_unit = PersistenceUnit('C:/testing/')
    actual = persistence_unit.save_results(trade_date_str, trade_time_str, aggregated_data, trade_exceptions_df,
                                           data_quality_summary)

    aggregated_data.to_csv.assert_called_once_with('C:/testing/PowerPosition_20220801_0000.csv', index=False)
    trade_exceptions_df.to_csv.assert_called_once_with('C:/testing/PowerPosition_20220801_0000_data_profiling.csv',
                                                       index=False)
    data_quality_summary.to_csv.assert_called_once_with('C:/testing/PowerPosition_20220801_0000_data_quality.csv')


src/powerservice/trading.py
"""
Module that contains the get trades functionality. This module will generate a random set of dummy positions.
"""
import random
import logging
import datetime
import uuid

import numpy as np
import pandas as pd


def check_if_valid_date(date: str):
    """
    Verify that the date format matches d/m/y
    :param date: str date in d/m/y format
    :return: True or False
    """
    date_format = "%d/%m/%Y"

    """ Warning to any non python devs reading this code..
        In Python the only way to test a valid date is with a try catch. Yep, it sux.
    """
    if not isinstance(date, str):
        return False

    try:
        datetime.datetime.strptime(date, date_format)
        valid_date = True
    except ValueError:
        valid_date = False

    return valid_date


def random_nan(x):
    """
    Replace x with a nan, if the random number == 1
    """
    if random.randrange(0, 15) == 1:
        x = np.nan

    return x


def generate_new_random_trade_position(date: str):
    """ Generates a new random trade position with the date, period sequence and volume sequence
    :param date: Date in d/m/y format
    :return: dict with data
    """

    period_list = [random_nan(i.strftime("%H:%M")) for i in pd.date_range("00:00", "23:59", freq="5min").time]
    volume = [random_nan(x) for x in random.sample(range(0, 500), len(period_list))]

    open_trade_position = {"date": date,
                           "time": period_list,
                           "volume": volume,
                           "id": uuid.uuid4().hex
                           }

    return open_trade_position


def get_trades(date: str):
    """
    Generate some random number of open trade positions
    :param date: date in d/m/y format
    :return:
    """

    if not check_if_valid_date(date=date):
        error_msg = "The supplied date {} is invalid.Please supply a date in the format d/m/Y.".format(date)
        logging.error(error_msg)
        raise ValueError(error_msg)

    # a randomly chosen number of open trades
    number_of_open_trades = random.randint(1, 101)
    logging.info("Generated" + str(number_of_open_trades) + " open trades randomly.")
    print("Generated" + str(number_of_open_trades) + " open trades randomly.")

    open_trades_list = []
    # Generate a list of open trade dicts
    for open_trade in range(0, number_of_open_trades):
        open_trades_list.append(generate_new_random_trade_position(date=date))

    return open_trades_list


if __name__ == '__main__':
    trades = get_trades(date='01/03/2022')
    df = pd.DataFrame(trades[0])
    print(df)



src/powerservice/client.py
import logging
import os
import time
from datetime import datetime

import pandas as pd

from powerservice.trading import get_trades, check_if_valid_date


class DataValidator:
    """
        This class implements an injectable data validator.
        The main method is validate(df), taking the input dataframe and returning a validated df
        it returns the validated dataframe with coulmns
        - invalid_time_format: true if timeformat is not "HH:MM"
        - unexpected_time: true if time is not on the 5mins mark
        - missing_time: if any record in the 24h/5min slot is missing
        - invalid_volume: if the volume is missing
    """

    def validate(self, df: pd.DataFrame) -> pd.DataFrame:
        """
        The map_reduce processing step expect the input data to be valid.
        This method orchestrates a battery of validators to
        mark valid and invalid data rows, so that they can be filtered out.
        :param df: The input dataframe for validation
        :return: The input dataframe augmented with the columns carrying the validation flags
        """
        # work on a copy of data to preserve the input dataframe from side effects
        validated_df = df.copy()
        validated_df = self._check_invalid_time_format(validated_df)
        validated_df = self._check_invalid_volume(validated_df)
        validated_df = self._check_unexpected_time(validated_df)
        validated_df = self._check_missing_time(validated_df)
        return validated_df

    def _check_invalid_time_format(self, validated_df: pd.DataFrame) -> pd.DataFrame:
        """
        The processor expect the time column to be formatted as HH:MM
        Therefore, we are marking all rows where volume is not populated or not parsing to the pattern
        :param validated_df: The DataFrame with the (partially) validated trades used for data quality profiling
        :return: The input dataframe augmented with invalid_time_format column marking any invalid value in the time column
        """
        # check if time is valid (we'll need to use apply as time checking in pandas is not vectorized
        validated_df['invalid_time_format'] = ~validated_df['time'].apply(self._is_hh_mm_time)
        return validated_df

    @staticmethod
    def _check_invalid_volume(validated_df: pd.DataFrame) -> pd.DataFrame:
        """
        The processor expect the volume column to be numeric and populated
        NOTE: Given no specific requirement there is no hard condition of positivity implemented.
        Therefore, we are marking all rows where volume is not populated or not parsing to a number
        :param validated_df: The DataFrame with the (partially) validated trades used for data quality profiling
        :return: The input dataframe augmented with invalid_volume column marking any invalid value in the volume column
        """
        # check invalid or missing volume
        validated_df['invalid_volume'] = pd.to_numeric(validated_df['volume'], errors='coerce').isnull()
        return validated_df

    @staticmethod
    def _check_unexpected_time(validated_df: pd.DataFrame) -> pd.DataFrame:
        """
        The processor expect the input to be provided on a 5 minutes frequency starting on midnight
        Therefore, we are expecting one row for each 5 minute interval of the day.
        This method identifies any interval not aligned with the frequency frequency
        :param validated_df: The DataFrame with the (partially) validated trades used for data quality profiling
        :return: The input dataframe augmented with unexpected_time column marking any misaligned row
        """
        # check times not on 5min interval
        unexpected_time = pd.DataFrame(pd.date_range("00:00", "23:59", freq="5min"), columns=['time'])
        unexpected_time['time'] = unexpected_time['time'].dt.strftime("%H:%M")
        unexpected_time['unexpected_time'] = False
        validated_df = validated_df.merge(unexpected_time, how='left', on='time')
        validated_df['unexpected_time'] = validated_df['unexpected_time'].fillna(True)
        return validated_df

    @staticmethod
    def _check_missing_time(validated_df: pd.DataFrame) -> pd.DataFrame:
        """
        The processor expect the input to be provided on a 5 minutes frequency starting on midnight
        Therefore, we are expecting one row for each 5 minute interval of the day.
        This method identifies missing intervals
        :param validated_df: The DataFrame with the (partially) validated trades used for data quality profiling
        :return: The input dataframe augmented with missing_time column and a new row for each missing interval
        """
        # check missing records on expected 5 minutes timeslots
        missing_time = pd.DataFrame(pd.date_range("00:00", "23:59", freq="5min"), columns=['time'])
        missing_time['time'] = missing_time['time'].dt.strftime("%H:%M")
        missing_time['missing_time'] = True
        temp_df = missing_time.merge(validated_df, how='left', on='time')
        temp_df['missing_time'] = temp_df['id'].isnull()
        temp_df = temp_df[temp_df['id'].isnull()]
        validated_df = pd.concat([validated_df, temp_df])
        validated_df['missing_time'] = validated_df['missing_time'].fillna(False)
        validated_df.loc[:, ['date', 'id']] = validated_df.loc[:, ['date', 'id']].ffill()
        return validated_df

    @staticmethod
    def _is_hh_mm_time(time_string: str) -> bool:
        """
        Returns true is a string is in format HH:MM else false.
        Basically a wrapper around a try/except block as python does not have a native function
        :param time_string: The DataFrame with the validated trades used for data quality profiling
        :return: True if time_string is formatted as HH:MM, False otherwise
        """
        try:
            time.strptime(str(time_string), '%H:%M')
            return True
        except ValueError:
            return False

    @staticmethod
    def get_valid_trades(validated_trades: pd.DataFrame) -> pd.DataFrame:
        """
        Given a validated dataframe it returns rows not marked with some data quality issue
        :param validated_trades_df: The DataFrame with the validated trades used for data quality profiling
        :return: The dataframe with the rows without data quality issues
        """
        return validated_trades[~(
            validated_trades['missing_time'] |
            validated_trades['invalid_time_format'] |
            validated_trades['unexpected_time'] |
            validated_trades['invalid_volume']
        )]

    @staticmethod
    def get_trade_exceptions(validated_trades_df: pd.DataFrame) -> pd.DataFrame:
        """
        Given a validated dataframe it returns rows marked with some issue
        :param validated_trades_df: The DataFrame with the validated trades used for data quality profiling
        :return: The dataframe with the data quality exception
        """
        return validated_trades_df[(
            validated_trades_df['missing_time'] |
            validated_trades_df['invalid_time_format'] |
            validated_trades_df['unexpected_time'] |
            validated_trades_df['invalid_volume']
        )]

    def get_data_quality_summary(self, validated_trades_df: pd.DataFrame) -> pd.DataFrame:
        """
        Given a validated dataframe it returns the data quality statistics per trade id
        :param validated_trades_df: The DataFrame with the validated trades used for data quality profiling
        :return: The dataframe with the data quality summary

        """
        errors_stats_df = validated_trades_df.groupby('id').agg({
            'missing_time': 'sum',
            'invalid_time_format': 'sum',
            'unexpected_time': 'sum',
            'invalid_volume': 'sum',
        })
        errors_stats_df = errors_stats_df.add_suffix('_count')

        # We'll compute min and max time only for the valid records
        valid_trades_df = self.get_valid_trades(validated_trades_df)
        time_stats_df = valid_trades_df.groupby('id').agg({
            'time': ['min', 'max']
        })
        # Flatten multi-index
        time_stats_df.columns = time_stats_df.columns.map('_'.join)
        result = errors_stats_df.join(time_stats_df)
        return result


class MapReduce:
    def __init__(self, input_timezone='Europe/London', output_timezone='UTC'):
        self.input_timezone = input_timezone
        self.output_timezone = output_timezone

    def map_reduce(self, valid_trades_df: pd.DataFrame) -> pd.DataFrame:
        """
        Given a valid dataframe with trade volumes on a 5 min frequency,
        returns a dataframe aggregating the data on an hourly basis,
        remapping time from input time zone to output time zone

        :param valid_trades_df: The DataFrame with the validated trades used for aggregation
        :return: The dataframe with the trades aggregated on hourly basis
        """

        # Parsing date and time to create an index we can aggregate on
        valid_trades_df['timestamp'] = \
            pd.to_datetime(valid_trades_df.agg('{0[date]} {0[time]}'.format, axis=1), dayfirst=True) \
                .dt.tz_localize(self.input_timezone)
        valid_trades_df = valid_trades_df.set_index(['timestamp'])

        # Aggregating the data on an hourly basis
        aggregated_trades = valid_trades_df.groupby(pd.Grouper(freq='60Min'))['volume'].sum().reset_index()

        # Mapping timestamp to UTC and then to HH:MM
        aggregated_trades['Local Time'] = \
            aggregated_trades['timestamp'].dt.tz_convert(self.output_timezone).dt.strftime("%H:%M")
        aggregated_trades['Volume'] = aggregated_trades['volume']
        aggregated_trades = aggregated_trades[['Local Time', 'Volume']]
        return aggregated_trades


class PersistenceUnit:

    def __init__(self, output_path):
        self._output_path = output_path

    def save_results(self, trade_date_str: str, trade_time_str: str, aggregated_data: pd.DataFrame,
                     trade_exceptions_df: pd.DataFrame, data_quality_summary: pd.DataFrame):
        """
        The method saves the input dataframes in the configured path using a file format like
        PowerPosition_YYYYMMDD_HHMM.csv,
        PowerPosition_YYYYMMDD_HHMM_data_profiling.csv,
        PowerPosition_YYYYMMDD_HHMM_data_quality.csv respectively.
        trade_date_str is replaced for YYYYMMDD and trade_time_str is replaced for HHMM
        For sake of this exercise, the file pattern is hardcoded, not parametrized.
        :param trade_date_str: The trade date in string format YYYYMMDD
        :param trade_time_str: The trade time in string format HHMM
        :param aggregated_data: The dataframe with the trades aggregated on hourly basis
        :param trade_exceptions_df: The dataframe with the exceptions found by the validator
        :param data_quality_summary: The Data quality summary for
        :return: None
        """
        filename = f"PowerPosition_{trade_date_str}_{trade_time_str}.csv"
        logging.info(f"Saving aggregated results to {filename}")
        aggregated_data.to_csv(os.path.join(self._output_path, filename), index=False)

        filename = f"PowerPosition_{trade_date_str}_{trade_time_str}_data_profiling.csv"
        logging.info(f"Saving data quality profiling to {filename}")
        trade_exceptions_df.to_csv(os.path.join(self._output_path, filename), index=False)

        filename = f"PowerPosition_{trade_date_str}_{trade_time_str}_data_quality.csv"
        logging.info(f"Saving data quality summary to {filename}")
        data_quality_summary.to_csv(os.path.join(self._output_path, filename))


class PetroineosChallenge:
    def __init__(self, validator, map_reduce, persistence_unit):
        self._validator = validator
        self._map_reduce = map_reduce
        self._persistence_unit = persistence_unit

    def process(self, trades: list[dict], trade_date: str) -> tuple[pd.DataFrame]:
        """
        The method takes the raw input from teh API, validates it, profiles the data quality and persists the results.
        The process is achieved using a Bridge pattern and delegating the core fuctions to handlers injected in the constructor.
        This make the processor agnostic to the actual processing achieving loose coupling between the parts
        and making testing easier.

        :param trades: A list of dictionaries as provided by the Mock-API
        :param trade_date: The trade date used to fetch the data from the Mock-API
        :return: a copy of aggregated_data_df, trade_exceptions, data_quality_summary_df as persisted
        """
        if not check_if_valid_date(date=trade_date):
            error_msg = f"The supplied date {trade_date} is invalid.Please supply a date in the format d/m/Y."
            logging.error(error_msg)
            raise ValueError(error_msg)

        logging.info(f"Processing trade date {trade_date}. We found {len(trades)} trades.")
        trade_exceptions_list = []
        valid_trades_list = []
        validated_trades_list = []

        logging.info("Validating trades...")
        for trade in trades:
            trade_df = pd.DataFrame(trade)
            validated_trades = self._validator.validate(trade_df)
            validated_trades_list.append(validated_trades)
            valid_trades_list.append(self._validator.get_valid_trades(validated_trades))
            trade_exceptions_list.append(self._validator.get_trade_exceptions(validated_trades))

        valid_trades_df = pd.concat(valid_trades_list)
        trade_exceptions_df = pd.concat(trade_exceptions_list)
        validated_trades_df = pd.concat(validated_trades_list)
        data_quality_summary_df = self._validator.get_data_quality_summary(validated_trades_df)
        if trade_exceptions_df.shape[0] > 0:
            # if the number of DQ issues is above a threshold we could throw an error and abort
            logging.warning(f"we found {trade_exceptions_df.shape[0]} data quality issues in the input data. "
                            f"Those line will be discarded")

        logging.info("Aggregating data...")
        aggregated_data_df = self._map_reduce.map_reduce(valid_trades_df)

        # Save results
        logging.info("Saving results...")
        trade_time_str = "0000"
        trade_date_str = datetime.strftime(datetime.strptime(trade_date, "%d/%m/%Y"), "%Y%m%d")
        self._persistence_unit.save_results(trade_date_str, trade_time_str, aggregated_data_df, trade_exceptions_df,
                                            data_quality_summary_df)

        return aggregated_data_df, trade_exceptions_df, data_quality_summary_df


def main(trade_date: str, output_path: str):
    """
    This is a very simple method showing the invocation sequence and the dependency injection
    """
    # Extract data for provider
    trades = get_trades(date=trade_date)

    # Configure and instantiate the processor
    processor = PetroineosChallenge(DataValidator(), MapReduce(), PersistenceUnit(output_path))

    # process data
    processor.process(trades, trade_date)


if __name__ == '__main__':
    # those values can come from caller or from configuration depending on the runtime design decisions
    trade_date = '01/08/2022'
    output_path = 'c:/tmp'

    main(trade_date, output_path)
